### PR TITLE
Adds pAI Insertion Into Objects

### PR DESCRIPTION
**Pull request by Shadowmech88**
This doesn't actually affect the game, this is purely the addition of a framework. I figure maybe now that it exists people might be inspired to do their own things with it.

This allows a pAI to be inserted into any `/obj` with its `can_take_pai` var set to `TRUE`. While inside an object, various hotkeys as well as mouse clicking can be used to trigger individual procs on the object corresponding to which hotkey was pressed. Altogether a pAI could manage 12 different functions of the object without even needing to use a menu.

Whenever an integrated pAI clicks an object that isn't its host object, `on_integrated_pai_click()` is called on the host object. By default, this simply calls `attack_pai()`, but if the proc has been overridden pAIs can still manually call `attack_pai()` on machinery by using Ctrl+click.

### CODE REVIEW

**Comment by jknpj** [comment]
pAI BUFFS FINALLY

File: code/game/objects/objs.dm [hunk 1]
```diff
@@ -46,6 +50,48 @@ var/global/list/reagents_to_log = list(FUEL, PLASMA, PACID, SACID, AMUTATIONTOXI
...
+/obj/proc/attack_integrated_pai(mob/user as mob)
+	return
+
+/obj/verb/remove_pai()
+	set name = "Remove pAI"
+	set category = "Object"
+	set src in range(1)
+
+	var/mob/living/carbon/human/H = usr
+	if(!istype(H))
```
Comment (PJB3005) [inline code/game/objects/objs.dm]: Do a regular dexterity check instead.

File: code/game/objects/objs.dm [hunk 2]
```diff
@@ -46,6 +50,48 @@ var/global/list/reagents_to_log = list(FUEL, PLASMA, PACID, SACID, AMUTATIONTOXI
...
+		to_chat(usr, "You don't have the dexterity to do this!")
+		return
+	if (H.stat == DEAD)
```
Comment (PJB3005) [inline code/game/objects/objs.dm]: Check all of these, and the rest, with `M.incapacitated()`.

File: code/game/objects/objs.dm (hunk 1 above)
Comment (PJB3005) [reply code/game/objects/objs.dm]: You also need to check `Adjacent`

File: code/game/objects/objs.dm [hunk 3]
```diff
@@ -46,6 +50,48 @@ var/global/list/reagents_to_log = list(FUEL, PLASMA, PACID, SACID, AMUTATIONTOXI
...
+	verbs += /obj/verb/remove_pai
+
+/obj/attackby(obj/item/weapon/W as obj, mob/user as mob)
+	if(can_take_pai && istype(W, /obj/item/device/paicard))
+		if(user.drop_item(W))
+			to_chat(user, "You insert \the [W] into a slot in \the [src].")
+			install_pai(W)
+			playsound(src, 'sound/misc/cartridge_in.ogg', 25)
+
+/obj/proc/attack_integrated_pai(mob/user as mob)
```
Comment (PJB3005) [inline code/game/objects/objs.dm]: Again

File: code/game/objects/objs.dm [hunk 4]
```diff
@@ -46,6 +50,48 @@ var/global/list/reagents_to_log = list(FUEL, PLASMA, PACID, SACID, AMUTATIONTOXI
...
 /obj/item/proc/is_used_on(obj/O, mob/user)
 
+/obj/proc/install_pai(obj/item/device/paicard/P as obj)
+	if(!P || !istype(P))
+		return 0
+	P.forceMove(src)
+	integratedpai = P
+	verbs += /obj/verb/remove_pai
+
+/obj/attackby(obj/item/weapon/W as obj, mob/user as mob)
```
Comment (PJB3005) [inline code/game/objects/objs.dm]: Again

File: code/game/objects/objs.dm [hunk 5]
```diff
@@ -46,6 +50,48 @@ var/global/list/reagents_to_log = list(FUEL, PLASMA, PACID, SACID, AMUTATIONTOXI
 
 /obj/item/proc/is_used_on(obj/O, mob/user)
 
+/obj/proc/install_pai(obj/item/device/paicard/P as obj)
```
Comment (PJB3005) [inline code/game/objects/objs.dm]: Remove `as obj`, it does nothing.

**Review by PJB3005 — CHANGES_REQUESTED** [review]

**Comment by Ephx1** [comment]
Unless someone actually goes through with this, the only way to buff pAIs now would be to nerf borers.
I don't even know what machine interactions you would have for most things aside from being a literal switch to toggle something.

**Comment by jknpj** [comment]
>I don't even know what machine interactions you would have for most things aside from being a literal switch to toggle something.

Egun pAI as ghetto Lawgiver

**Comment by pvak** [comment]
pai inside a space suit

**Comment by Ephx1** [comment]
That is a genuinely cool feature from Bay that will probably never make its way here due to our hardsuits being simple clothing.
And seeing how it's only clothing, the only feature I _could_ see being implemented is the ability for the pAI to control the suit if the wearer dies or passes out.

**Comment by aceedex** [comment]
Maybe I'm stupid for thinking this, but you could recycle some oldborer code for it

**Comment by jknpj** [comment]
>That is a genuinely cool feature from Bay that will probably never make its way here due to our hardsuits being simple clothing.

They can even use carded AIs in their rigsuit, but then again, rigsuits are way too broken.

>I don't even know what machine interactions you would have for most things aside from being a literal switch to toggle something.

Maybe some items getting stronger alt-versions when activated pAI are inside them? <s>but that's being a literal switch to toggle something.

**Comment by Ephx1** [comment]
I actually did outline a few things a loooooong time ago you could put pAIs in, but they were all the simple bots. If you could do that, that would be pretty neat.
Put into medibot if you can open the cover, then you can have a pAI steering it, using whatever it has in its beaker. You could do much of the same for all the other bots, since they all have one function.

I also wanted pAIs to be able to move, but not use mechs, however from what I've seen mech code is already a godawful mess so no one will ever even think of doing that.

File: code/game/objects/objs.dm (hunk 1 above)
Comment (Shadowmech88) [reply code/game/objects/objs.dm]: Do any mobs besides humans and methylin monkeys have dexterity?
And isn't `Adjacent` checking handled by the verb's requirement for `range(1)`?

File: code/game/objects/objs.dm (hunk 5 above)
Comment (Shadowmech88) [reply code/game/objects/objs.dm]: For future reference, does it always do nothing?

Comment (PJB3005) [reply code/game/objects/objs.dm]: When using a verb it IS relevant because then BYOND knows what type the variable is.

Otherwise no.

**Comment by Kitfox88** [comment]
>stuff pAI into every mechanical limb
>become megazord

**Comment by Shadowmech88** [comment]
Hopefully with pAI buffs will come more pAI players to fill the expanded roles.

File: code/game/objects/objs.dm (hunk 1 above)
Comment (Exxion) [reply code/game/objects/objs.dm]: Adjacent() also checks for blockages.

**Comment by Shadowmech88** [comment]
* `/obj`s now also have a `on_integrated_pai_click()` proc, which is called whenever an integrated pAI clicks something that isn't the object itself. By default, the proc simply calls attack_pai().
* While integrated, pAIs can also manually attack_pai() machinery by using Ctrl+click.

This would allow, for example, a pAI to be placed inside a gun and operate it from within, firing at targets onscreen.

**Comment by Shadowmech88** [comment]
Since, as far as I know, pAIs have no current use for most hotkeys or for intents, I expanded the number of input-receiving `/obj` procs for integrated pAIs.

* Integrated pAIs can now manipulate their host objects via the equip, drop, throw, swap, cycle-intent-right, cycle-intent-left, help, disarm, grab, and harm hotkeys, as well as the `attack_self()` hotkey and clicking.

This means that a pAI could feasibly control 12 different functions of an object without having to even open a menu.

**Comment by yclatious** [comment]
Please allow them to use * actions when inside stuff, I am sure sec would very much appreciate having a stun baton or egun spam *LAW as they beat the shit out of people or stun them.

**Comment by jknpj** [comment]
>I also wanted pAIs to be able to move, but not use mechs, however from what I've seen mech code is already a godawful mess so no one will ever even think of doing that.

Being a co-pilot (switching equips, managing battery, lights, radio, et cetera) would be feasible, no? Even being able to eject your master's dead body from inside a mech would be enogh motivation to carry one with you.

**Comment by Shadowmech88** [comment]
>Please allow them to use * actions when inside stuff

I haven't specifically prohibited it, but I'm not sure how it works when a mob uses an emote while in the contents of another object.
Can pAIs even emote at all, as-is?

**Comment by yclatious** [comment]
Im fairly certain they cannot, they cant even inspect stuff whilst in pockets.

**Comment by Shadowmech88** [comment]
>they cant even inspect stuff whilst in pockets

This PR should actually fix their inspecting, since their overridden `ClickOn()` proc now properly calls `ShiftClickOn()` if they're Shift+clicking.

**Review by PJB3005 — APPROVED** [review]

**Comment by PJB3005** [comment]
Is this done?

**Comment by Shadowmech88** [comment]
Yes.

**Comment by Shadowmech88** [comment]
Now the real fun can begin.

**Comment by Probe1** [comment]
pls no pAI power creep

**Comment by Intigracy** [comment]
zhmbqtx

**Pull request merged**